### PR TITLE
[SPARK-34533][SQL] Eliminate LEFT ANTI join to empty relation in AQE

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

I discovered from review discussion - https://github.com/apache/spark/pull/31630#discussion_r581774000 , that we can eliminate LEFT ANTI join (with no join condition) to empty relation, if the right side is known to be non-empty. So with AQE, this is doable similar to https://github.com/apache/spark/pull/29484 .

### Why are the changes needed?

This can help eliminate the join operator during logical plan optimization.
Before this PR, [left side physical plan `execute()` will be called](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala#L192), so if left side is complicated (e.g. contain broadcast exchange operator), then some computation would happen. However after this PR, the join operator will be removed during logical plan, and nothing will be computed from left side. Potentially it can save resource for these kinds of query.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit tests for positive and negative queries in `AdaptiveQueryExecSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
cc @cloud-fan and @maropu to take a look if you have time, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala [hunk 1]
```diff
@@ -53,5 +57,23 @@ object EliminateJoinToEmptyRelation extends Rule[LogicalPlan] {
 
     case j @ Join(_, _, LeftSemi, _, _) if canEliminate(j.right, EmptyHashedRelation) =>
       LocalRelation(j.output, data = Seq.empty, isStreaming = j.isStreaming)
+
+    case j @ Join(_, _, LeftAnti, None, _) =>
+      val isNonEmptyBroadcastedRightSide = j.right match {
+        case LogicalQueryStage(_, stage: BroadcastQueryStageExec)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala:63]: nit:
```
case LogicalQueryStage(_, stage: QueryStageExec) =>
  stage.getRuntimeStatistics.rowCount.exists(_ > 0)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala [hunk 2]
```diff
@@ -33,6 +34,9 @@ import org.apache.spark.sql.execution.joins.{EmptyHashedRelation, HashedRelation
  *    This applies to all Joins (sort merge join, shuffled hash join, and broadcast hash join),
  *    because sort merge join and shuffled hash join will be changed to broadcast hash join with AQE
  *    at the first place.
+ *
+ * 3. Join is left anti join without condition, and broadcasted join right side is not empty.
+ *    This applies to broadcast nested loop join only.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala:39]: Why do we care about the physical plan? I think it's a logical optimization that left anti join without condition can be turned into empty relation if right side is non-empty.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala]: @cloud-fan - thanks, use row count is better, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala]: @cloud-fan - I agree. Updated to remove this restriction. I was not thinking towards to use row count stats at the first place, so was checking the size of `Array[InternalRow]` for `BroadcastNestedLoopJoinExec`, so was having this restriction.

**Comment by c21** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 3]
```diff
@@ -1230,6 +1237,30 @@ class AdaptiveQueryExecSuite
...
+          // broadcast non-empty right side
+          ("SELECT /*+ broadcast(testData3) */ * FROM testData LEFT ANTI JOIN testData3", true),
+          // broadcast empty right side
+          ("SELECT /*+ broadcast(emptyTestData) */ * FROM testData LEFT ANTI JOIN emptyTestData",
+            false),
+          // broadcast left side
+          ("SELECT /*+ broadcast(testData) */ * FROM testData LEFT ANTI JOIN testData3", false)
+        ).foreach { case (query, isEliminated) =>
+          val (plan, adaptivePlan) = runAdaptiveAndVerifyResult(query)
+          val bnlj = findTopLevelBroadcastNestedLoopJoin(plan)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1255]: Do we need to create `findTopLevelBroadcastNestedLoopJoin`? We just want to prove that, the non-AQE plan has join   and the AQE one has no join. `findTopLevelBaseJoin` should be good enough here.

Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: @cloud-fan - agree, updated.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by maropu** [comment]
Thanks for the update, @c21 !

**Comment by c21** [comment]
Thank you @cloud-fan and @maropu for review!

**Pull request closed without merging**